reference-implementation/lib/abstract-ops/readable-streams.js

```diff
@@ -1456,6 +1456,6 @@ function ReadableStreamFromIterable(asyncIterable) {
     return promiseResolvedWith(returnResult);
   }
 
-  stream = CreateReadableStream(startAlgorithm, pullAlgorithm, cancelAlgorithm);
+  stream = CreateReadableStream(startAlgorithm, pullAlgorithm, cancelAlgorithm, 0);
```

Comment:
Setting the high-water mark to zero ensures `ReadableStream.from(otherReadableStream)` doesn't immediately start pulling from `otherReadableStream`, and matches the default HWM for `TransformStream.readable` (step 5 in [the `TransformStream` constructor](https://streams.spec.whatwg.org/commit-snapshots/660d259fa263e3c0646415a41b6e7b9a13ac4d04/#ts-constructor)).

In comparison, `ReadableStream.tee()` *does* immediately start pulling, even if no `read()` was called on any of the returned branches. In step 16 and 17 of [the `ReadableStreamTee` abstract op](https://streams.spec.whatwg.org/commit-snapshots/660d259fa263e3c0646415a41b6e7b9a13ac4d04/#readable-stream-tee), no value is passed for *highWaterMark* and thus it defaults to 1. This seems to be intentional, as there are tests that rely on this. (In particular, we rely on it to forward closing of the parent stream to the two branches.)

I *think* we want to follow the example of `TransformStream.readable` here, and use zero as HWM? Or am I missing something important?

Reply:
Hmm. I have a slight bias toward keeping our default, since it must have been picked for a reason :). In particular, I believe our historical intention has been that by default readable streams should try to keep something in their internal queue, whether that something is derived from a `controller.enqueue()` call in `start()`, or one in `pull()`. I think that same logic applies to deriving it from an async iterator.

Transform streams are a bit different, as we want introducing them to be more of a no-op and not cause a bunch of chunks to sit in intermediate transform-stream queues, unless explicitly requested.

On the other hand, you could view `ReadableStream.from()` as a sort of transform, such that the same no-op logic should apply. Hmm. @ricea, any thoughts?

Reply:
Oh right, I didn't update the spec text yet to pass a zero high-water mark. But I'll first wait until we decide what that high-water mark should be. 😉

> Transform streams are a bit different, as we want introducing them to be more of a no-op and not cause a bunch of chunks to sit in intermediate transform-stream queues, unless explicitly requested.

Hmm... it doesn't look like that's working as intended? 😕 

By default, a `TransformStream` is indeed created with a readable HWM of 0, but it still has a **writable HWM** of 1. So piping a stream into it *will* pull at least one chunk, to fill the writable end's queue.
```javascript
var rs1 = new ReadableStream({
  pull(c) {
    console.log('pull');
    c.enqueue('a');
  }
}, { highWaterMark: 0 });
// no "pull" is logged yet
var rs2 = rs1.pipeThrough(new TransformStream());
// logs "pull" once
```

There's also no good way to fix this. You can't set both the readable and writable HWM to 0, since then the pipe will stall:
```javascript
var rs1 = new ReadableStream({
  pull(c) {
    console.log('pull');
    c.enqueue('a');
  }
}, { highWaterMark: 0 });
var rs2 = rs.pipeThrough(new TransformStream({}, { highWaterMark: 0 }));
var r = rs.getReader();
await r.read(); // never resolves
```

Reply:
It feels like there's a connection missing somewhere... There's no way for the writable end of a transform stream to know that the readable end is being read from. 🤔 
```javascript
let { readable, writable } = new TransformStream({}, { highWaterMark: 0 });
let r = readable.getReader();
let w = writable.getWriter();

// we start reading from the readable end...
let read1 = r.read();

// ...but the writable end is still applying backpressure. :-/
w.desiredSize; // still 0
w.ready; // still pending
```

Reply:
> By default, a `TransformStream` is indeed created with a readable HWM of 0, but it still has a **writable HWM** of 1. So piping a stream into it _will_ pull at least one chunk, to fill the writable end's queue.

Yes. As you observed, a writable stream with a HWM of 0 will always have backpressure. So adding an identity TransformStream to a pipe can't be a complete no-op: it always increases the total queue size by 1.

> Hmm. I have a slight bias toward keeping our default, since it must have been picked for a reason :).

In implementation practice we seem to be setting readable HWM to 0 whenever we create a platform stream, because it permits maximum control over backpressure. So I'm not sure the default of 1 is correct.

Reply:
Another reason for setting HWM to 0: you can treat `ReadableStream.from()` as the "inverse" of `ReadableStream.prototype.values()`.
```javascript
const it1 = ['a', 'b', 'c'];
// This async iterator should behave the same as the original
const it2 = ReadableStream.from(it1).values();
for await (const chunk of it2) { /* ... */ }
```
```javascript
const rs1 = new ReadableStream({
  start(c) {
    c.enqueue('a');
    c.enqueue('b');
    c.enqueue('c');
    c.close();
  }
});
// This stream should behave the same as the original
const rs2 = ReadableStream.from(rs1.values());
const reader = rs2.getReader();
for (let result = await reader.read(); !result.done; result = await reader.read()) { /* ... */ }
```